rfc/20250317-ephemeral-resources.md

```diff
@@ -0,0 +1,646 @@
+* `RenewAt` being an optional timestamp indicating when OpenTofu will have to call `Renew` method on the provider before using again the data from the `Result`.
+
+Observations:
+* In the `Result`, OpenTofu is epecting to find any non-computed given values in the request, otherwise will return an error.
+* In the `Result`, the fields marked as computed can be either null or have an actual value. If an unknown if found, OpenTofu will return an error.
+
+> [!NOTE]
+>
+> If any information in the configuration of an ephemeral resource is unknown during the `plan` phase, OpenTofu will defer the provisioning of the resource for the `apply` phase.
+
+#### `Renew` method details
+The `Renew` method is called only if the response from `Open` or another `Renew` call is containing a `RenewAt`.
+When `RenewAt` is present, OpenTofu, before using the `Result` from the `Open` method response, will check if the current timestamp is at or over `RenewAt` and will call the `Renew` method by providing the previously returned `Private` information.
+
+> [!NOTE]
+> 
+> `Renew` does not return a *new* information meant to replace the initial `Result` returned by the `Open` call.
+> Due to this, `Renew` is only useful for system similar to Vault where the lease can be renewed without generating new data.
+
+#### `Close` method details
+When OpenTofu is done using an ephemeral resource, it needs to call its `Close` method to ensure that any remote data associated with the data returned in `OpenResponse.Result` is released and/or cleaned up properly.
+The `Renew` request is requiring the latest `Private` data returned by the call to `Open` or `Renew` method.
+
+#### `ConfigValidators` and `ValidateConfig` methods details
+There is not much to say here, since this is the same lifecycle that a datasource is having.
+
+#### Testing support
+The testing support will be documented later into a different RFC, or as amendment to this one.
+
+### Support in already ephemeral contexts
+There are already OpenTofu contexts that are not saved in state/plan file:
+* `provider` configuration
+* `provisioner` blocks
+* `connection` blocks
+
+In all of these, referencing an ephemeral value should work as normal.
+
+### Utilities
+#### `terraform.applying`
+The `terraform.applying` needs to be introduced to allow the user to check if the current command that is running is `apply` or not.
+This is useful when user wants to configure different properties between write operations and read operations.
+
+`terraform.applying` will be set to `true` when `tofu apply` is executed and `false` in any other command.
+
+> [!NOTE]
+>
+> This keyword is related to the `apply` command and not to the `apply` phase, meaning that when running `tofu apply`, `terraform.applying` will still be `true` also during the `plan` phase of the `apply` command.
```

Comment:
what about during destroy?

Reply:
Fixed in 2c6587907eae9c404931ce789091dfd192f5628b